### DISCUSSION TITLE
Can we connect a new social auth after the user is logged in?

### DISCUSSION THREAD

**Discussion by RatulSaha** [supabase/supabase#6679] [Questions]
This is similar to https://github.com/supabase/supabase/discussions/5827, but not completely.

The use case is: The user logs in with email or a social like Twitter. Then,
1. we want to check that she has not logged in with her discord,
2. Have a "Connect Discord" button in their profile page so that they can connect their discord as well.

The reason: We want to provide some discord specific use cases for those who want it, for which we want to know what is the Discord details of a particular user. But we can't just only Discord as an auth.

**Comment by kangmingtay** [comment]
Hey @RatulSaha, for (1) you can check if the user has / has not logged in with discord by checking the `providers` field in the `auth.users.raw_app_meta_data` column. As for (2), you can add the "connect discord button" to the profile page and if the user's discord primary email is the same as the user's twitter email, then gotrue will automatically link the identities to the same user.

Since you have a constraint that states that you can't use discord only as an auth provider, you can add an after insert trigger that checks if the discord identity created is linked to the logged-in user. If the user's discord email is different from their twitter email, then gotrue will create a new user and identity in `auth.users` and `auth.identities` respectively. You can then check the newly inserted `auth.users.raw_app_meta_data->>'providers'` field again and reject the signup in the trigger accordingly.

**Reply by chhuang** [reply]
Hi @kangmingtay, could you elaborate a bit more on the topic please?

We are having an issue where users are trying to connect to a social account that has a different email address, this causes Supabase to create a new account for them and switches the user to that new account.

**Example:**

1. I join using my Facebook, my Facebook email is me@abc.com
2. After I logged in with Facebook, I try to add/connect my Twitter, but my Twitter email is me@xyz.com

**What actual happens:**
Supabase will create a new account for me@xyz.com and auto log me in to that account

**What I would like to happen:**
Before Supabase goes and creates a new account, I would like to compare the existing account (me@abc.com) vs the one from my Twitter (me@xyz.com). 

If the emails are different, I want to instruct Supabase so it donesn't create the new account but returns a warning message: "The twitter account you are trying to connect has a different email than me@abc.com, please try updating your twitter's email or try with a different twitter account."

If the emails are the same, proceed as normal (Supabase will add Twitter to the existing account with the same email).

Is there anyway I can intercept/add some conditions to `signInWithOAuth()` before Supabase goes and creates a new account? 

Something like:

```js
if (signInWithOAuth('twitter').email !== my_current_email) don't_create_a_new_account
```

**Reply by kangmingtay** [reply]
hi @chhuang,

> We are having an issue where users are trying to connect to a social account that has a different email address

This is currently not supported yet but we have plans to provide APIs to enable manual linking of accounts. 

Currently, we only support automatic linking of accounts which is why the emails have to be the same. Automatic linking relies on the assumption that if a new user has the same email as an existing user **and** the email has been verified, the new user should be merged with the existing user. Because of this assumption, if the email returned from the oauth provider hasn't been seen before, it will always be treated as a new user.

> Is there anyway I can intercept/add some conditions to signInWithOAuth() before Supabase goes and creates a new account?

Unfortunately, we don't have a way to do this right now.

**Comment by charnowsky** [comment]
> hi @chhuang,
> 
> > We are having an issue where users are trying to connect to a social account that has a different email address
> 
> This is currently not supported yet but we have plans to provide APIs to enable manual linking of accounts.
> 
> Currently, we only support automatic linking of accounts which is why the emails have to be the same. Automatic linking relies on the assumption that if a new user has the same email as an existing user **and** the email has been verified, the new user should be merged with the existing user. Because of this assumption, if the email returned from the oauth provider hasn't been seen before, it will always be treated as a new user.
> 
> > Is there anyway I can intercept/add some conditions to signInWithOAuth() before Supabase goes and creates a new account?
> 
> Unfortunately, we don't have a way to do this right now.

Hi, is there any update on that?

**Reply by GaryAustin1** [reply]
https://github.com/supabase/gotrue/issues/313  looks like where it is being "tracked".
